SimG4Core/Application/plugins/OscarMTProducer.h

```diff
@@ -38,6 +49,13 @@ class OscarMTProducer : public edm::stream::EDProducer<edm::GlobalCache<OscarMTM
 private:
   std::unique_ptr<RunManagerMTWorker> m_runManagerWorker;
   const OscarMTMasterThread* m_masterThread = nullptr;
+
+  static edm::ESGetToken<cms::DDCompactView, IdealGeometryRecord> m_DD4Hep;
```

Comment:
Instead of being static, a new global cache object should be made which holds the OscarMTMasterThread and these tokens. It would require the tokens to be marked mutable so they can be set in the constructor.

Reply:
@Dr15Jones, situation is not simple - I started from non-static but compilation fail, because  the method is static:

static std::shared_ptr<int> globalBeginRun(const edm::Run& iRun, .....)

static method cannot use non-static class members.

Reply:
Tokens are moved to the master thread.